### PR TITLE
ZeRO-1/2: add bounded quantized weight all-gather

### PR DESCRIPTION
**Pull request by adenzhou1350**
## Summary

Extend the existing `zero_quantized_weights` opt-in to ZeRO stages 1 and 2. After each optimizer step, weight partitions are quantized in bounded chunks, gathered, and dequantized one source rank at a time into the persistent flat buffer.

Large weights use 64-element symmetric INT8 groups. Parameters with at most 65,536 elements use a bounded original-dtype side channel; they are masked before quantization and restored after gathering. The bucket limit applies across all data-parallel ranks, with a minimum of one quantization group per rank. Unsupported FP32 and ZenFlow combinations are rejected during optimizer initialization.

The patch also removes an unused `cuda_profiler_api.h` include from `gemm_test.h`; CUDA 13 no longer ships that header and the quantizer does not use the profiler API. The default ZeRO-1/2 path is unchanged.

## Validation (`f2c781d`)

- Changed-file pre-commit checks passed.
- Full focused test file: **28 passed, 0 failed/skipped** — 22 CPU cases (including simulated DP64 buffer bounds) and 6 real two-GPU RTX 5090 cases. The GPU cases cover ZeRO-1/2 FP16/BF16 training and public gather reconstruction across chunk/partition boundaries. The quantizer was freshly built from this revision.
- Qwen3-4B-Base, sequence 2048, BF16, ZeRO-2 DP4 on four PCIe RTX 5090 GPUs: four fresh-process arms in baseline/candidate/candidate/baseline order, each with 1 warmup + 19 measured steps.
- Within-arm loss/gradient trajectories and sampled parameter hashes reproduce exactly across orders. Cross-arm maximum loss difference: 0.004051; 1,194 parameter samples: relative L2 0.000460, max absolute difference 0.0078125, zero `atol=0.02, rtol=0.02` violations.
- Whole training-step speedup: **1.08515x**; within-run 95% bootstrap CI **[1.08418, 1.08590]**. Baseline medians: 1.19319 / 1.19339 s; candidate: 1.09943 / 1.09989 s.
- Peak allocated/reserved memory: 28.88 / 30.25 GB in both arms, within the unchanged 90% / 95% gates.

The DP4 run used PyTorch 2.11.0+cu130, CUDA 13.0, NCCL CTA8, a 26M reduce bucket and a 50M all-gather bucket. Both arms used the same exact source; only `zero_quantized_weights` differed.

## Scope and risks

This is lossy, opt-in synchronization. The short BF16 Qwen run is not a convergence study; its within-run interval does not establish cross-day or cross-machine robustness. FP16 has focused DP2 correctness coverage, not whole-model performance qualification. MoE, offload, other models/accelerators, NVLink/RDMA, and long-horizon convergence are not established. Small-parameter-heavy workloads may not benefit from the extra side-channel collectives.

Prepared and tested with Codex assistance.

### CODE REVIEW

File: deepspeed/runtime/utils.py [hunk 1]
```diff
@@ -1024,6 +1024,130 @@ def all_gather_into_tensor_dp_groups(groups_flat, partitioned_param_groups, dp_p
...
+                                            device=local_partition.device)
+
+        for offset in range(0, local_partition.numel(), chunk_elements):
+            count = min(chunk_elements, local_partition.numel() - offset)
+            padded_elements = (
+                (count + quantization_group_size - 1) // quantization_group_size) * quantization_group_size
+            groups = padded_elements // quantization_group_size
+            padded[:count].copy_(local_partition.narrow(0, offset, count))
+            padded[count:padded_elements].zero_()
+            quantized, scales = quantizer.quantize(padded[:padded_elements], groups=groups)
```
[the remark this answers is not in the document]
Comment (adenzhou1350) [reply deepspeed/runtime/utils.py:1154]: Fixed in `c0debf7`: every preserved intersection is zeroed in the padded quantizer input before scale calculation, while the gathered original-dtype values are restored afterward. The focused test covers a large preserved value sharing a group with small neighboring weights.

File: deepspeed/runtime/utils.py [hunk 2]
```diff
@@ -1024,6 +1024,130 @@ def all_gather_into_tensor_dp_groups(groups_flat, partitioned_param_groups, dp_p
...
+                                                                             dp_world_size)
+        if max_preserved_count:
+            local_preserved = torch.zeros(max_preserved_count,
+                                          dtype=local_partition.dtype,
+                                          device=local_partition.device)
+            for local_offset, count, packed_offset in preserved_layouts[partition_id]:
+                local_preserved.narrow(0, packed_offset, count).copy_(local_partition.narrow(0, local_offset, count))
+            gathered_preserved = torch.empty(dp_world_size * max_preserved_count,
+                                             dtype=local_partition.dtype,
+                                             device=local_partition.device)
```
[the remark this answers is not in the document]
Comment (adenzhou1350) [reply deepspeed/runtime/utils.py]: Fixed in `c0debf7`: the original-dtype side channel now iterates bounded chunks under the same bucket limit instead of allocating and gathering a full flattened group. Tests cover chunking across parameter and partition boundaries.

File: tests/unit/runtime/zero/test_quantized_weight_allgather.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    quantizer = FakeQuantizer()
+    monkeypatch.setattr(utils.dist, "get_rank", lambda group=None: 0)
+    monkeypatch.setattr(utils.dist, "get_world_size", lambda group=None: 2)
+
+    def fake_all_gather(output, source, group=None):
+        output[:source.numel()].copy_(source)
+        output[source.numel():2 * source.numel()].copy_(source)
+
+    monkeypatch.setattr(utils.dist, "all_gather_into_tensor", fake_all_gather)
+    utils.all_gather_quantized_dp_groups([group_flat], [partitions], [object()], 6, quantizer, 8)
```
[the remark this answers is not in the document]
Comment (adenzhou1350) [reply tests/unit/runtime/zero/test_quantized_weight_allgather.py:46]: A device-backed ZeRO-1/2 integration test is now committed in `c0debf7`. It passes locally on an RTX 4060 with the real CUDA quantizer extension and actual optimizer steps, but world size one bypasses the collective path. Exact-head DP2/DP4 execution is still pending, so I am leaving this thread unresolved and keeping the PR in draft.

[the remark this answers is not in the document]
Comment (adenzhou1350) [reply tests/unit/runtime/zero/test_quantized_weight_allgather.py:50]: Ran the committed tests on two RTX 5090s at c0debf7: ZeRO-1 and ZeRO-2 both passed, with no skips. The CUDA quantizer was freshly built from this head, and the run exercised NCCL and optimizer steps. Updated the PR description; the older DP4 performance evidence remains labeled historical.

**Comment by adenzhou1350** [comment]
Addressed all four findings in `f2c781d`: early FP32/ZenFlow rejection, aggregate all-ranks buffer limits, and public reconstruction tests instead of private-layout assertions.

Fresh-head validation: 28/28 focused tests pass, including six real DP2 cases; the same-condition, order-balanced DP4 Qwen run passes correctness/memory gates at 1.085x speedup. The PR body now has the updated results and their scope.
